lib/ServiceManager.ts

```diff
@@ -0,0 +1,158 @@
+
+export class ServiceManager implements ServiceManagerInterface {
+  private sdk: OktaAuth;
+  private options: ServiceManagerOptions;
+  private services: Map<string, ServiceInterface>;
+  private channel?: BroadcastChannel;
+  private elector?: LeaderElector;
+  private started: boolean;
+
+  private static knownServices = ['autoRenew', 'syncStorage'];
+
+  private static defaultOptions = {
+    autoRenew: true,
+    autoRemove: true,
+    syncStorage: true
+  };
+
+  constructor(sdk: OktaAuth, options: ServiceManagerOptions = {}) {
+    this.sdk = sdk;
+
+    // TODO: backwards compatibility, remove in next major version - OKTA-473815
+    const { autoRenew, autoRemove, syncStorage } = sdk.tokenManager.getOptions();
+    this.options = Object.assign({}, 
+      ServiceManager.defaultOptions,
+      { autoRenew, autoRemove, syncStorage },
+      options
+    );
+
+    this.started = false;
+    this.services = new Map();
+    this.onLeaderDuplicate = this.onLeaderDuplicate.bind(this);
+    this.onLeader = this.onLeader.bind(this);
+  }
+
+  public static canUseLeaderElection() {
+    return isBrowser();
+  }
+
+  private onLeader() {
+    if (this.started) {
+      // Start services that requires leadership
+      this.startServices();
+    }
+  }
+
+  private onLeaderDuplicate() {
+  }
+
+  isLeader() {
+    return !!this.elector?.isLeader;
+  }
+
+  hasLeader() {
+    return this.elector?.hasLeader;
+  }
+
+  start() {
+    if (this.started) {
+      this.stop();
+    }
+    this.startElector();
+    this.startServices();
+    this.started = true;
+  }
+  
+  stop() {
+    this.stopElector();
+    this.stopServices();
+    this.started = false;
+  }
+
+  getService(name: string): ServiceInterface | undefined {
+    return this.services.get(name);
+  }
+
+  private startServices() {
+    for (const name of ServiceManager.knownServices) {
+      const srv = this.createService(name);
+      if (srv) {
+        const canStart = srv.canStart() && !srv.isStarted() && (srv.requiresLeadership() ? this.isLeader() : true);
+        if (canStart) {
+          srv.start();
+          this.services.set(name, srv);
+        }
+      }
+    }
+  }
+
+  private stopServices() {
+    for (const srv of this.services.values()) {
+      srv.stop();
+    }
+    this.services = new Map();
+  }
+
+  private startElector() {
+    if (ServiceManager.canUseLeaderElection()) {
+      if (!this.channel) {
+        const { broadcastChannelName } = this.options;
+        this.channel = new BroadcastChannel(broadcastChannelName as string);
+      }
+      if (!this.elector) {
+        this.elector = createLeaderElection(this.channel);
+        this.elector.onduplicate = this.onLeaderDuplicate;
+        this.elector.awaitLeadership().then(this.onLeader);
+      }
+    }
+  }
+
+  private stopElector() {
+    this.elector?.die();
+    this.elector = undefined;
+  }
+
+  private createService(name: string): ServiceInterface | undefined {
```

Comment:
We should be able to remove `undefined ` from the type, as it throws instead of returning `undefined`.